systick.cpp

```diff
@@ -110,23 +156,126 @@ unsigned long t_systick1 = 0;
 unsigned long t_systick2 = 0;
 unsigned long t_systick3 = 0;
 
-// The systick event is an ISR attached to Timer 2
-// This currently runs at 2ms or 500Hz.
-ISR(TIMER2_COMPA_vect) {
-  //unsigned long _start = micros();
-  updateBatteryVolts();
-  //t_systick1 = micros() - _start;
-
-  //_start = micros();
-  updateFunctionSwitch();
-  //t_systick2 = micros() - _start;
 
-  //_start = micros();
-  //updateWallSensor();
-  update_sensors_control();
-  //t_systick3 = micros() - _start;
-
-  //speed_control();
+/***
+ * This is the systick event - an ISR connected to Timer 2
+ * It currently runs at 20kHz so that the emitter on duty
+ * is low.
+ * There are 40 possible time-slots, or phases, if the systick
+ * is intended to be synchronous with the main control loop
+ * running every 2 milliseconds. 
+ * Each tick is called every 50us so that an analogue conversion 
+ * started in one phase will be completed ready for the next phase. 
+ * This technique avoids having the processor sitting idle
+ * during conversions and while waiting for the sensors to respond.
+ * So that high-speed encoder interrupts are not missed each phase
+ * should ideally last no more than 10us.
+ * 
+ * The reason fo rthis apparently arcane technique is that the 
+ * ATMEGA328P does not have nested interrupts. At top speed, the 
+ * encoders may generate interrupt at more than 20kHz and we cannot
+ * afford to miss one. Also, received serial characters must be serviced
+ * as they arrive or they will be lost. Having a single systick event
+ * that does everything but takes many tens of microseconds risks
+ * lost serial data and encoder pulses.
+ *  
+ * For testing, the built-in LED is turned on at the start of the interrupt
+ * and off again at the end. This can be used to measure system load.
+ * Disable the feature if you want to use the built-in LED for some other 
+ * purpose.
+ */ 
+ISR(TIMER2_COMPA_vect) {
+  digitalWriteFast(LED_BUILTIN, 1);
```

Comment:
Is this a testing line?

Reply:
I can remove the LED flashing. It is just for timing with an oscilloscope.